Everywhere: Make all headers compile again

BenWiederhake opened the pull request:
Did you know that the following does not compile?

```c++
#include <AK/RecursionDecision.h>
```

But it should! With this PR, all the library headers compile again.

This PR:
- Cleans up badly-formatted includes
- Cleans up an overly strict rule in HeaderCheck (foreign-arch headers cannot be always compiled)
- Adds in missing headers
- Moves an overly-inline function out of line
- Adds in even more missing headers

I handled LibWeb separately because LibWeb has some funky inter-header dependencies. This should make it easier to bisect if something did indeed go wrong.

konradekk commented on Meta/HeaderCheck/generate_all.py:
```diff
@@ -49,6 +49,7 @@ def generate_part(header):
 def run(root_path, arch):
     os.chdir(root_path)
     headers_list = get_headers_here()
+    headers_list = [h for h in headers_list if not h.startswith("Userland/Libraries/LibC/sys/arch/aarch64")]
```
(does it deserve a `FIXME` or `TODO`?)

BenWiederhake replied:
You're right!

BenWiederhake commented:
Changes since initial push:
- Added a FIXME in case someone ever wants to run HeaderCheck on other architectures
- Rebased for convenience

linusg reviewed (changes requested):
Why is there both `Everywhere: Include required headers` and `LibWeb: Include required headers`?

trflynn89 commented on Userland/Libraries/LibVideo/Track.h:
```diff
@@ -11,6 +11,7 @@
 #include <AK/Traits.h>
 #include <AK/Types.h>
 #include <AK/Variant.h>
+#include <AK/kmalloc.h>
```
Cursory glance - I don't see any symbol from kmalloc.h being used here?

BenWiederhake replied:
Variant.h uses placement new in some methods, which are not always generated. Importing it in Variant.h causes lots of new inclusions, but here it's actually triggered.

trflynn89 replied:
Yeah I disagree with this approach then. If Variant.h needs a header, it should include it itself, it doesn't make sense for a random LibVideo header to have to resolve Variant's non-forward-declarable dependencies for it.

What header does Variant.h *actually* need? Because it doesn't use any kmalloc.h symbols either. If the error was due to placement new, I'd assume `<new>`, for which we have existing examples on how to keep `<new>` out of the kernel (e.g. Singleton.h).

BenWiederhake replied:
Putting it into Variant.h is fine with me.

Trying to include `<new>` is indeed more direct, and causes the exact problems that you suggested. However, implementing any `#ifdef`-logic ad-hoc in Variant.h feels like duplicating the `#ifdef`s in kmalloc.h, which seems to be a bad idea.

Would including kmalloc.h in Variant.h work for you? This seems to be the closest practical solution to what you suggested.

BenWiederhake replied:
Actually, it's less complicated logic than I expected:
```c++
#if defined(KERNEL)
#    include <Kernel/Heap/kmalloc.h>
#else
#    include <new>
#endif
```

It still seems weird to me, I'd prefer just putting `<kmalloc.h>` in Variant.h

BenWiederhake commented:
Because LibWeb has some special inter-header dependencies. In the past there were always some special problems, like header inclusion cycles, that made changes like this more difficult. And in fact, there still are such cycles. To make it easier to write and debug, I split it across two commits.

<details><summary>The cycles, in long form.</summary>

```
digraph G {
  // Cycle or component around Userland/Libraries/LibC/sys/types.h
  "Userland/Libraries/LibC/sys/types.h" -> "Kernel/API/POSIX/sys/types.h";
  "Kernel/API/POSIX/sys/types.h" -> "Userland/Libraries/LibC/sys/types.h";
  // Cycle or component around AK/Array.h
  "AK/Array.h" -> "AK/Span.h";
  "AK/Span.h" -> "AK/Array.h";
  // Cycle or component around Kernel/Process.h
  "Kernel/Process.h" -> "Kernel/Jail.h";
  "Kernel/Jail.h" -> "Kernel/Process.h";
  // Cycle or component around Userland/Libraries/LibGfx/Rect.h
  "Userland/Libraries/LibGfx/Rect.h" -> "Userland/Libraries/LibGfx/Line.h";
  "Userland/Libraries/LibGfx/Line.h" -> "Userland/Libraries/LibGfx/Rect.h";
  // Cycle or component around Kernel/Bus/USB/USBDevice.h
  "Kernel/Bus/USB/USBDevice.h" -> "Kernel/Bus/USB/USBConfiguration.h";
  "Kernel/Bus/USB/USBConfiguration.h" -> "Kernel/Bus/USB/USBDevice.h";
  // Cycle or component around Kernel/Storage/ATA/AHCI/InterruptHandler.h
  "Kernel/Storage/ATA/AHCI/InterruptHandler.h" -> "Kernel/Storage/ATA/AHCI/Port.h";
  "Kernel/Storage/ATA/AHCI/Port.h" -> "Kernel/Storage/ATA/AHCI/InterruptHandler.h";
  // Cycle or component around Userland/Libraries/LibWeb/DOM/AccessibilityTreeNode.h
  "Userland/Libraries/LibWeb/DOM/AccessibilityTreeNode.h" -> "Userland/Libraries/LibWeb/DOM/Node.h";
  "Userland/Libraries/LibWeb/DOM/Node.h" -> "Userland/Libraries/LibWeb/DOM/AccessibilityTreeNode.h";
  // Cycle or component around Userland/Services/AudioServer/Mixer.h
  "Userland/Services/AudioServer/Mixer.h" -> "Userland/Services/AudioServer/FadingProperty.h";
  "Userland/Services/AudioServer/FadingProperty.h" -> "Userland/Services/AudioServer/Mixer.h";
  // Cycle or component around Userland/Services/WebContent/ConnectionFromClient.h
  "Userland/Services/WebContent/ConnectionFromClient.h" -> "Userland/Services/WebContent/WebContentConsoleClient.h";
  "Userland/Services/WebContent/WebContentConsoleClient.h" -> "Userland/Services/WebContent/ConnectionFromClient.h";
}
```

</details>

Are you just asking for information, or was that a suggestion to squash these two commits?

BenWiederhake commented:
Changes since last push:
- Rebased for convenience
- Fixed a logic mistake: In generate_all.py, the function `is_foreign_to` had a `not` too many, producing always the exactly wrong result.
- Squashed the "Include required headers" commits as suggested by linusg.

BenWiederhake commented:
Changes since last push:
- Rebased for convenience
- Use the same trick as in Singleton.h to make Variant always work, see also https://github.com/SerenityOS/serenity/pull/18262#discussion_r1161341454

BenWiederhake commented:
It seems like noone cares whether our headers compile or not :(
